File: services/tokocrypto.py
#!/usr/bin/env python3
"""
===============================================
TOKOCRYPTO API INTEGRATION - PYTHON VERSION
MAHA LAKSHMI HOLDINGS
===============================================

Python alternative for environments without Node.js

Usage:
    python3 tokocrypto.py 20000

Requirements:
    pip install requests python-dotenv

"""

import os
import sys
import json
import hmac
import hashlib
import time
import requests
from datetime import datetime

try:
    from dotenv import load_dotenv
except ImportError:
    print("Installing python-dotenv...")
    os.system("pip install python-dotenv")
    from dotenv import load_dotenv

# ================================================
# CONFIGURATION
# ================================================

# Load .env file
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
ENV_FILE = os.path.join(SCRIPT_DIR, ".env.payout")

if os.path.exists(ENV_FILE):
    load_dotenv(ENV_FILE)
    print("✅ Loaded .env.payout configuration")

# Tokocrypto credentials
API_KEY = os.getenv("TOKOCRYPTO_API_KEY", "")
API_SECRET = os.getenv("TOKOCRYPTO_API_SECRET", "")
BTC_WALLET = os.getenv("BTC_WALLET_ADDRESS", "1H3FZkKsX6jgTuqA23fduLVtxL7MrtgWe2")
SIMULATION_MODE = os.getenv("SIMULATION_MODE", "false").lower() == "true"

# API Configuration
BASE_URL = "https://www.tokocrypto.com"
API_VERSION = "/open/v1"

# ================================================
# HELPER FUNCTIONS
# ================================================

def create_signature(query_string: str) -> str:
    """Generate HMAC SHA256 signature"""
    return hmac.new(
        API_SECRET.encode('utf-8'),
        query_string.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()


def api_request(method: str, endpoint: str, params: dict = None, signed: bool = False) -> dict:
    """Make API request to Tokocrypto"""
    timestamp = int(time.time() * 1000)
    
    # Build query parameters
    query_params = [f"timestamp={timestamp}"]
    if params:
        for key, value in params.items():
            query_params.append(f"{key}={value}")
    
    query_string = "&".join(query_params)
    
    # Add signature for signed endpoints
    if signed and API_SECRET and API_SECRET != "YOUR_API_SECRET_HERE":
        signature = create_signature(query_string)
        query_params.append(f"signature={signature}")
        query_string = "&".join(query_params)
    
    url = f"{BASE_URL}{API_VERSION}{endpoint}?{query_string}"
    
    headers = {
        "X-MBX-APIKEY": API_KEY,
        "Content-Type": "application/json"
    }
    
    print(f"\n🔗 API Request: {method} {endpoint}")
    
    try:
        if method == "GET":
            response = requests.get(url, headers=headers, timeout=30)
        elif method == "POST":
            response = requests.post(url, headers=headers, timeout=30)
        else:
            raise ValueError(f"Unsupported method: {method}")
        
        return response.json()
    
    except requests.exceptions.RequestException as e:
        return {"success": False, "error": str(e)}


def get_btc_price() -> dict:
    """Get current BTC/IDR price"""
    result = api_request("GET", "/market/ticker", {"symbol": "BTC_IDR"})
    
    if result.get("data") and len(result["data"]) > 0:
        return {
            "price": float(result["data"][0].get("lastPrice", 0)),
            "high": float(result["data"][0].get("highPrice", 0)),
            "low": float(result["data"][0].get("lowPrice", 0))
        }
    
    # Return default if API fails
    return {"price": 89300000, "source": "DEFAULT"}


def place_market_order(side: str, quote_qty: float) -> dict:
    """Place market buy/sell order"""
    params = {
        "symbol": "BTC_IDR",
        "symbolType": 3,
        "side": side.upper(),
        "type": 1,  # MARKET order
        "quoteQty": str(quote_qty)
    }
    
    return api_request("POST", "/orders", params, signed=True)


def withdraw_btc(amount: float, address: str, network: str = "BTC") -> dict:
    """Withdraw BTC to address"""
    params = {
        "amount": str(amount),
        "coin": "BTC",
        "address": address,
        "network": network
    }
    
    return api_request("POST", "/asset/withdraw", params, signed=True)


# ================================================
# MAIN EXECUTION
# ================================================

def execute_payout(idr_amount: int) -> dict:
    """Execute CEO payout - buy BTC and withdraw to wallet"""
    
    print("\n" + "=" * 70)
    print("🏦 MAHA LAKSHMI - CEO PAYOUT EXECUTION (Python)")
    print("=" * 70)
    print(f"📅 Timestamp: {datetime.now().isoformat()}")
    print(f"💰 Amount: Rp {idr_amount:,}")
    print(f"🎯 Mode: {'SIMULATION' if SIMULATION_MODE else 'LIVE'}")
    print(f"👛 BTC Wallet: {BTC_WALLET}")
    print("=" * 70)
    
    # Check if API secret is configured
    has_api_secret = API_SECRET and API_SECRET != "YOUR_API_SECRET_HERE"
    
    if SIMULATION_MODE or not has_api_secret:
        print("\n⚠️ SIMULATION MODE - No actual transaction will be made")
        print(f"   Reason: {'Simulation mode enabled' if SIMULATION_MODE else 'API_SECRET not configured'}")
        
        btc_price = 89300000
        btc_amount = idr_amount / btc_price
        
        return {
            "success": True,
            "mode": "SIMULATION",
            "steps": {
                "step1_buy_btc": {
                    "action": "Market Buy BTC",
                    "idr_amount": idr_amount,
                    "btc_amount": round(btc_amount, 8),
                    "btc_price": btc_price,
                    "status": "SIMULATED",
                    "txid": f"SIM-TX-{int(time.time()*1000)}"
                },
                "step2_withdraw": {
                    "action": "Withdraw BTC",
                    "btc_amount": round(btc_amount, 8),
                    "destination": BTC_WALLET,
                    "network": "BTC",
                    "status": "SIMULATED",
                    "txid": f"SIM-WD-{int(time.time()*1000)}"
                }
            },
            "total_btc": round(btc_amount, 8),
            "destination_wallet": BTC_WALLET,
            "message": "⚠️ SIMULATION MODE - Configure API_SECRET for live execution"
        }
    
    # LIVE EXECUTION
    try:
        # Step 1: Get BTC price
        print("\n📊 Step 1: Getting BTC price...")
        price_info = get_btc_price()
        btc_price = price_info["price"]
        print(f"💵 Current BTC price: Rp {btc_price:,}")
        
        # Step 2: Calculate BTC amount
        btc_amount = idr_amount / btc_price
        print(f"💰 BTC amount: {btc_amount:.8f} BTC")
        
        # Step 3: Place buy order
        print("\n📊 Step 2: Placing market buy order...")
        buy_result = place_market_order("BUY", idr_amount)
        
        if not buy_result.get("success", False) and buy_result.get("code") != 0:
            raise Exception(f"Buy order failed: {json.dumps(buy_result)}")
        
        order_id = buy_result.get("data", {}).get("orderId")
        print(f"✅ Buy order placed! Order ID: {order_id}")
        
        # Step 4: Withdraw BTC
        print("\n📊 Step 3: Withdrawing BTC to CEO wallet...")
        withdraw_result = withdraw_btc(btc_amount, BTC_WALLET)
        
        if not withdraw_result.get("success", False) and withdraw_result.get("code") != 0:
            raise Exception(f"Withdrawal failed: {json.dumps(withdraw_result)}")
        
        txid = withdraw_result.get("data", {}).get("txId") or withdraw_result.get("data", {}).get("txid")
        print(f"✅ Withdrawal initiated! TxID: {txid}")
        
        return {
            "success": True,
            "mode": "LIVE",
            "steps": {
                "step1_get_price": {
                    "action": "Get BTC Price",
                    "price": btc_price,
                    "status": "COMPLETED"
                },
                "step2_buy_btc": {
                    "action": "Market Buy BTC",
                    "order_id": order_id,
                    "idr_amount": idr_amount,
                    "btc_amount": round(btc_amount, 8),
                    "status": "COMPLETED"
                },
                "step3_withdraw": {
                    "action": "Withdraw BTC",
                    "txid": txid,
                    "btc_amount": round(btc_amount, 8),
                    "destination": BTC_WALLET,
                    "network": "BTC",
                    "status": "PENDING_CONFIRMATION"
                }
            },
            "total_btc": round(btc_amount, 8),
            "destination_wallet": BTC_WALLET,
            "message": "✅ Live execution completed!"
        }
        
    except Exception as e:
        print(f"\n❌ EXECUTION FAILED: {str(e)}")
        return {
            "success": False,
            "error": str(e)
        }


# ================================================
# CLI ENTRY POINT
# ================================================

if __name__ == "__main__":
    # Get amount from command line
    if len(sys.argv) > 1:
        try:
            amount = int(sys.argv[1])
        except ValueError:
            print(f"❌ Error: Invalid amount '{sys.argv[1]}'")
            print("Usage: python3 tokocrypto.py <amount_in_IDR>")
            sys.exit(1)
    else:
        amount = 20000  # Default test amount
    
    print(f"\n🎯 Test Amount: Rp {amount:,}")
    
    # Execute payout
    result = execute_payout(amount)
    
    # Print result
    print("\n📋 Result:")
    print(json.dumps(result, indent=2))
    
    # Save to log file
    log_file = os.path.join(SCRIPT_DIR, f"execution-log-{datetime.now().strftime('%Y%m%d-%H%M%S')}.json")
    with open(log_file, 'w') as f:
        json.dump(result, f, indent=2)
    print(f"\n📁 Log saved to: {log_file}")
    
    sys.exit(0 if result.get("success") else 1)

File: services/tokocrypto.js
/**
 * ================================================
 * TOKOCRYPTO API INTEGRATION
 * MAHA LAKSHMI HOLDINGS
 * ================================================
 * 
 * Integration for BTC Market Order & Withdrawal
 * API Documentation: https://www.tokocrypto.com/apidocs
 * 
 * VERSION: 1.0.0
 * DATE: 2026-07-17
 * 
 * SECURITY:
 * - API keys stored in environment variables
 * - HMAC SHA256 signature for all signed requests
 * - IP whitelist required (34.10.175.217)
 * ================================================
 */

const https = require('https');
const crypto = require('crypto');

// ================================================
// CONFIGURATION
// ================================================

const CONFIG = {
    baseUrl: 'https://www.tokocrypto.com',
    apiVersion: '/open/v1',
    
    // Load from environment
    apiKey: process.env.TOKOCRYPTO_API_KEY,
    apiSecret: process.env.TOKOCRYPTO_API_SECRET,
    
    // Trading pair
    tradingPair: 'BTC_IDR', // For BTC/IDR trading
    symbolType: 3, // IDR trading pair
    
    // BTC withdrawal address
    btcWallet: process.env.BTC_WALLET_ADDRESS || '1H3FZkKsX6jgTuqA23fduLVtxL7MrtgWe2',
    
    // Simulation mode
    simulationMode: process.env.SIMULATION_MODE !== 'false'
};

// ================================================
// SIGNATURE GENERATOR (HMAC SHA256)
// ================================================

function createSignature(queryString, secret) {
    // HMAC SHA256 signature - hex lowercase
    return crypto
        .createHmac('sha256', secret)
        .update(queryString)
        .digest('hex');
}

// ================================================
// API REQUEST HANDLER
// ================================================

function apiRequest(method, endpoint, params = {}, signed = false) {
    return new Promise((resolve, reject) => {
        // Build query string with sorted parameters
        const timestamp = Date.now();
        let queryParams = `timestamp=${timestamp}`;
        
        // Sort parameters alphabetically for signature
        if (Object.keys(params).length > 0) {
            const sortedKeys = Object.keys(params).sort();
            for (const key of sortedKeys) {
                queryParams += `&${key}=${params[key]}`;
            }
        }
        
        // Create signature if signed endpoint
        let signature = '';
        if (signed && CONFIG.apiSecret && CONFIG.apiSecret !== 'YOUR_API_SECRET_HERE') {
            signature = createSignature(queryParams, CONFIG.apiSecret);
            queryParams += `&signature=${signature}`;
        }
        
        // Debug: log what we're signing
        console.log(`   🔐 Signing query: ${queryParams.substring(0, 100)}...`);
        
        const path = `${CONFIG.apiVersion}${endpoint}?${queryParams}`;
        
        const options = {
            hostname: 'www.tokocrypto.com',
            path: path,
            method: method,
            headers: {
                'X-MBX-APIKEY': CONFIG.apiKey || '',
                'Content-Type': 'application/json'
            }
        };
        
        console.log(`\n🔗 API Request: ${method} ${endpoint}`);
        
        const req = https.request(options, (res) => {
            let data = '';
            
            res.on('data', chunk => {
                data += chunk;
            });
            
            res.on('end', () => {
                try {
                    const result = JSON.parse(data);
                    resolve(result);
                } catch (error) {
                    resolve(data);
                }
            });
        });
        
        req.on('error', (error) => {
            reject(error);
        });
        
        req.end();
    });
}

// ================================================
// ACCOUNT INFO
// ================================================

async function getAccountInfo() {
    console.log('\n📋 Fetching account info...');
    const result = await apiRequest('GET', '/account/accountInfo', {}, true);
    return result;
}

// ================================================
// GET BTC PRICE
// ================================================

async function getBTCPrice() {
    console.log('\n💰 Fetching BTC price...');
    const result = await apiRequest('GET', '/market/ticker', { 
        symbol: CONFIG.tradingPair 
    });
    
    // Try different response formats
    if (result.data && result.data.length > 0) {
        return {
            price: parseFloat(result.data[0].lastPrice),
            highPrice: parseFloat(result.data[0].highPrice),
            lowPrice: parseFloat(result.data[0].lowPrice),
            volume: parseFloat(result.data[0].volume)
        };
    } else if (result.lastPrice) {
        // Direct format
        return {
            price: parseFloat(result.lastPrice),
            highPrice: parseFloat(result.highPrice),
            lowPrice: parseFloat(result.lowPrice),
            volume: parseFloat(result.volume)
        };
    } else if (result.data && result.data.lastPrice) {
        // Nested data format
        return {
            price: parseFloat(result.data.lastPrice),
            highPrice: parseFloat(result.data.highPrice),
            lowPrice: parseFloat(result.data.lowPrice),
            volume: parseFloat(result.data.volume)
        };
    }
    
    // Return default price for simulation
    console.log('⚠️ Could not fetch live price, using default');
    return {
        price: 89300000, // Default BTC price in IDR
        source: 'DEFAULT'
    };
}

// ================================================
// PLACE MARKET ORDER (BUY BTC)
// ================================================

async function placeMarketOrder(side, quoteQty) {
    console.log(`\n📝 Placing ${side} order for Rp ${quoteQty.toLocaleString('id-ID')}...`);
    
    const params = {
        symbol: CONFIG.tradingPair,
        symbolType: CONFIG.symbolType,
        side: side, // BUY or SELL
        type: 1, // MARKET order
        quoteQty: quoteQty.toString()
    };
    
    const result = await apiRequest('POST', '/orders', params, true);
    
    return {
        success: result.success || result.code === 0,
        orderId: result.data?.orderId,
        clientId: result.data?.clientId,
        executedQty: result.data?.executedQty,
        executedQuoteQty: result.data?.executedQuoteQty,
        status: result.data?.status,
        raw: result
    };
}

// ================================================
// WITHDRAW BTC
// ================================================

async function withdrawBTC(amount, address, network = 'BTC') {
    console.log(`\n🚀 Withdrawing ${amount} BTC to ${address}...`);
    
    const params = {
        amount: amount.toString(),
        coin: 'BTC',
        address: address,
        network: network // BTC, BEP20, TRC20, etc.
    };
    
    const result = await apiRequest('POST', '/asset/withdraw', params, true);
    
    return {
        success: result.success || result.code === 0,
        withdrawId: result.data?.id || result.data?.withdrawId,
        txid: result.data?.txId || result.data?.txid,
        status: result.data?.status,
        raw: result
    };
}

// ================================================
// GET DEPOSIT/WITHDRAW HISTORY
// ================================================

async function getWithdrawHistory(limit = 10) {
    console.log('\n📜 Fetching withdraw history...');
    const result = await apiRequest('GET', '/asset/withdraw/history', { 
        coin: 'BTC',
        limit: limit 
    }, true);
    return result;
}

// ================================================
// BUY BTC & WITHDRAW TO CEO WALLET
// ================================================

async function executeCEOPayout(idrAmount) {
    const startTime = Date.now();
    
    console.log('\n' + '='.repeat(70));
    console.log('🏦 MAHA LAKSHMI - CEO PAYOUT EXECUTION');
    console.log('='.repeat(70));
    console.log(`📅 Timestamp: ${new Date().toISOString()}`);
    console.log(`💰 Amount: Rp ${idrAmount.toLocaleString('id-ID')}`);
    console.log(`🎯 Mode: ${CONFIG.simulationMode ? 'SIMULATION' : 'LIVE'}`);
    console.log(`👛 BTC Wallet: ${CONFIG.btcWallet}`);
    console.log('='.repeat(70));
    
    // Check if API secret is configured
    const hasApiSecret = CONFIG.apiSecret && CONFIG.apiSecret !== 'YOUR_API_SECRET_HERE';
    
    if (CONFIG.simulationMode || !hasApiSecret) {
        console.log('\n⚠️ SIMULATION MODE - No actual transaction will be made');
        console.log(`   Reason: ${CONFIG.simulationMode ? 'Simulation mode enabled' : 'API_SECRET not configured'}`);
        console.log('');
        
        // Get simulated BTC price
        const btcPrice = 89300000; // Simulated price
        const btcAmount = idrAmount / btcPrice;
        
        return {
            success: true,
            mode: 'SIMULATION',
            steps: {
                step1_buy_btc: {
                    action: 'Market Buy BTC',
                    idr_amount: idrAmount,
                    btc_amount: btcAmount.toFixed(8),
                    btc_price: btcPrice,
                    status: 'SIMULATED',
                    txid: 'SIM-TX-' + Date.now()
                },
                step2_withdraw: {
                    action: 'Withdraw BTC',
                    btc_amount: btcAmount.toFixed(8),
                    destination: CONFIG.btcWallet,
                    network: 'BTC',
                    status: 'SIMULATED',
                    txid: 'SIM-WD-' + Date.now()
                }
            },
            total_btc: btcAmount.toFixed(8),
            destination_wallet: CONFIG.btcWallet,
            execution_time_ms: Date.now() - startTime,
            message: '⚠️ SIMULATION MODE - Configure API_SECRET and set SIMULATION_MODE=false for live execution'
        };
    }
    
    try {
        // Step 1: Get current BTC price
        console.log('\n📊 Step 1: Getting BTC price...');
        const priceInfo = await getBTCPrice();
        const btcPrice = priceInfo.price;
        console.log(`💵 Current BTC price: Rp ${btcPrice.toLocaleString('id-ID')}`);
        
        // Step 2: Calculate BTC amount
        const btcAmount = idrAmount / btcPrice;
        console.log(`💰 BTC amount: ${btcAmount.toFixed(8)} BTC`);
        
        // Step 3: Place market buy order
        console.log('\n📊 Step 2: Placing market buy order...');
        const buyResult = await placeMarketOrder('BUY', idrAmount);
        
        if (!buyResult.success) {
            throw new Error(`Buy order failed: ${JSON.stringify(buyResult)}`);
        }
        
        console.log(`✅ Buy order placed! Order ID: ${buyResult.orderId}`);
        
        // Step 4: Withdraw BTC to CEO wallet
        console.log('\n📊 Step 3: Withdrawing BTC to CEO wallet...');
        const withdrawResult = await withdrawBTC(btcAmount, CONFIG.btcWallet);
        
        if (!withdrawResult.success) {
            throw new Error(`Withdrawal failed: ${JSON.stringify(withdrawResult)}`);
        }
        
        console.log(`✅ Withdrawal initiated! TxID: ${withdrawResult.txid}`);
        
        const executionTime = Date.now() - startTime;
        
        console.log('\n' + '='.repeat(70));
        console.log('✅ EXECUTION COMPLETE');
        console.log('='.repeat(70));
        
        return {
            success: true,
            mode: 'LIVE',
            steps: {
                step1_get_price: {
                    action: 'Get BTC Price',
                    price: btcPrice,
                    status: 'COMPLETED'
                },
                step2_buy_btc: {
                    action: 'Market Buy BTC',
                    order_id: buyResult.orderId,
                    client_id: buyResult.clientId,
                    idr_amount: idrAmount,
                    btc_amount: btcAmount.toFixed(8),
                    status: 'COMPLETED'
                },
                step3_withdraw: {
                    action: 'Withdraw BTC',
                    withdraw_id: withdrawResult.withdrawId,
                    txid: withdrawResult.txid,
                    btc_amount: btcAmount.toFixed(8),
                    destination: CONFIG.btcWallet,
                    network: 'BTC',
                    status: 'PENDING_CONFIRMATION'
                }
            },
            total_btc: btcAmount.toFixed(8),
            destination_wallet: CONFIG.btcWallet,
            execution_time_ms: executionTime,
            message: '✅ Live execution completed. Withdrawal pending blockchain confirmation.'
        };
        
    } catch (error) {
        console.error('\n❌ EXECUTION FAILED:', error.message);
        
        return {
            success: false,
            error: error.message,
            execution_time_ms: Date.now() - startTime
        };
    }
}

// ================================================
// EXPORT MODULES
// ================================================

module.exports = {
    CONFIG,
    getAccountInfo,
    getBTCPrice,
    placeMarketOrder,
    withdrawBTC,
    getWithdrawHistory,
    executeCEOPayout
};

// ================================================
// CLI EXECUTION
// ================================================

if (require.main === module) {
    (async () => {
        // Load .env if exists
        const fs = require('fs');
        const path = require('path');
        
        const envPath = path.join(__dirname, '.env.payout');
        if (fs.existsSync(envPath)) {
            const envContent = fs.readFileSync(envPath, 'utf8');
            envContent.split('\n').forEach(line => {
                const [key, ...valueParts] = line.split('=');
                if (key && valueParts.length > 0 && !key.startsWith('#')) {
                    process.env[key.trim()] = valueParts.join('=').trim();
                }
            });
            console.log('✅ Loaded .env.payout configuration');
        }
        
        // Update config with loaded values
        CONFIG.apiKey = process.env.TOKOCRYPTO_API_KEY;
        CONFIG.apiSecret = process.env.TOKOCRYPTO_API_SECRET;
        CONFIG.btcWallet = process.env.BTC_WALLET_ADDRESS || CONFIG.btcWallet;
        CONFIG.simulationMode = process.env.SIMULATION_MODE === 'true';
        
        // Get test amount from command line or use default
        const testAmount = parseInt(process.argv[2]) || 20000; // Default Rp 20,000
        
        console.log(`\n🎯 Test Amount: Rp ${testAmount.toLocaleString('id-ID')}`);
        
        // Execute payout
        const result = await executeCEOPayout(testAmount);
        
        console.log('\n📋 Result:');
        console.log(JSON.stringify(result, null, 2));
        
        process.exit(result.success ? 0 : 1);
    })();
}
